[17.0][IMP] fs_storage: replace {db_name} in directory_path, fs_attachment and fs_attachment_s3: adapt to change

vincent-hatakeyama opened the pull request:
In multi database mode, there was no way to avoid filename collisions.

In multi database mode, without using that feature in fs_attachment, as the created filename does not contain the database name, it is possible to overwrite files from another database if obfuscation is off and the file have the same name and the same id or cause deletion of files if the checksum is used as the name and the same file is copied over several database and unlinked in one of them.
I would have prefer for the name of the file to always include the database name but I think it would have been a major change considered too breaking. Allowing to use the database name seemed more compatible.

lmignon commented:
@vincent-hatakeyama you should use one backend by database. IMO this change is useless.

sbidoul reviewed (changes requested):
The use case is unclear to me and this could indeed be solved with a different backend configuration per database.

I can imagine this could facilitate the automatic creation of that backend.

This would need a test, and an update of the help text of the field to reveal the feature to users.

sbidoul commented on fs_storage/models/fs_storage.py:
```diff
@@ -162,6 +162,13 @@ def __init__(self, env, ids=(), prefetch_ids=()):
+        # replace {db_name} with the dbname
+        if field_name == "directory_path":
+            value = value.format(db_name=self.env.cr.dbname)
+        return value
```
This is also probably not the best place to implement this, as if we do this, it should not behave differently when the value is entered manually or via server environment.

vincent-hatakeyama replied:
Reworked as indicated in https://github.com/OCA/storage/pull/379#issuecomment-2970362090

vincent-hatakeyama commented:
To clarify the use case:

- Odoo is used in multi-tenant mode. New database can be created by some users.
- fs_storage and fs_attachment are used
- the configuration to store the files is stored in odoo configuration file.
- The files referenced by a database should not be deleted when deleted in another base.

With those point in mind, I thought that allowing db_name and changing its value would allow splitting the file per database.

> you should use one backend by database. IMO this change is useless.

In the case I’ve described, this is not possible. To my knowledge, it is not possible to indicate to server_environment to use db_name in any of the fields.

sbidoul commented:
Thanks for the clarification. This makes sense.

adrienpeiffer commented:
@vincent-hatakeyama  Thanks for that
Indeed, it makes sense. However, as Stéphane said, it would be preferable to put this in another place so that it's also possible when the value is set manually.

vincent-hatakeyama commented:
I plan to change how this is handled when I have the time and to add tests to cover the case.

khalidchamakh commented:
hello @vincent-hatakeyama, 
Any news about your PR? We're having the same problem.

vincent-hatakeyama commented:
The current version in the PR is working and is in running in productions (for at least a year).

I haven’t had the time yet to rework the PR.

Lionel-cl commented:
I've been working on this topic. I'll post my proposal today in a new PR.

vincent-hatakeyama committed:
[IMP] fs_storage: replace {db_name} in directory_path, fs_attachment and fs_attachment_s3: adapt to change

In multi database mode, there was no way to avoid filename collisions
when a configuration file is used to define the fs_storage.storage to
store attachments.

vincent-hatakeyama commented:
I’ve reworked the PR.

Substitution of `{db_name}` is now done in all cases.
I choose to introduce a get_directory_path method to handle the substitution. That seemed easier than introducing another field and make directory_path computed while also configurable with server_environment.
While looking at it, I found out that adaptations are also needed in fs_attachment.

Side note, the module uses {db_name} to be similar with the substitution done in web_ribbon.

lmignon commented on fs_storage/readme/USAGE.md:
```diff
@@ -75,6 +75,8 @@ for the following fields:
 - options
 - directory_path
 
+When evaluating directory_path, `{db_name}` is replaced by the database name.
```
@vincent-hatakeyama Can you explain the use case where it could be a good practice to use the `{db_name}` placeholder in your directory path?

vincent-hatakeyama replied:
I’ve added a sentence to indicate when the feature is to be used and rebased the branch.

lmignon reviewed (approved):
Thank you @vincent-hatakeyama for your contribution, your patience and your tenacity so that we understand the need ;-)

Lionel-cl commented:
Thanks for that. My PR is then no longer necessary. 
However, this code (although more conventional) also requires changing all modules dependent on fs_storage.

vincent-hatakeyama commented:
@lmignon Any chance this gets merged as the port to 18.0 (#514) has been merged?

lmignon reviewed (changes requested):
@vincent-hatakeyama  A fix seems to be missing for `fs_attachment_s3` https://github.com/OCA/storage/pull/514/files#diff-f384ce2e99b19f6cb9bd99fa6a15fa155805bcfb90bf5861fc90a146305fdf19

lmignon commented:
/ocabot merge minor
